MGDSTRM-1664 Resource support for latency metrics

slaskawi opened the pull request:
## Motivation

https://issues.redhat.com/browse/MGDSTRM-1664

In order to support better observability, all the performance metrics need to support resources. This way, we know exactly which endpoints are fine and which cause some problems.

## What

This Pull Request introduces a new "resource" field into the latency metrics. Here's an example:

```
keycloak_request_duration_bucket{method="POST",resource="/protocol/openid-connect/token",le="50.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="/protocol/openid-connect/token",le="100.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="/protocol/openid-connect/token",le="250.0",} 2.0
keycloak_request_duration_bucket{method="POST",resource="/protocol/openid-connect/token",le="500.0",} 2.0
keycloak_request_duration_bucket{method="POST",resource="/protocol/openid-connect/token",le="1000.0",} 2.0
keycloak_request_duration_bucket{method="POST",resource="/protocol/openid-connect/token",le="2000.0",} 2.0
keycloak_request_duration_bucket{method="POST",resource="/protocol/openid-connect/token",le="10000.0",} 2.0
keycloak_request_duration_bucket{method="POST",resource="/protocol/openid-connect/token",le="30000.0",} 2.0
keycloak_request_duration_bucket{method="POST",resource="/protocol/openid-connect/token",le="+Inf",} 2.0
keycloak_request_duration_count{method="POST",resource="/protocol/openid-connect/token",} 2.0
keycloak_request_duration_sum{method="POST",resource="/protocol/openid-connect/token",} 122.0
keycloak_request_duration_bucket{method="GET",resource="/admin",le="50.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin",le="100.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin",le="250.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin",le="500.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin",le="1000.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin",le="2000.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin",le="10000.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin",le="30000.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin",le="+Inf",} 14.0
keycloak_request_duration_count{method="GET",resource="/admin",} 14.0
keycloak_request_duration_sum{method="GET",resource="/admin",} 39.0
keycloak_request_duration_bucket{method="POST",resource="/login-actions/authenticate",le="50.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="/login-actions/authenticate",le="100.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="/login-actions/authenticate",le="250.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="/login-actions/authenticate",le="500.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="/login-actions/authenticate",le="1000.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="/login-actions/authenticate",le="2000.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="/login-actions/authenticate",le="10000.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="/login-actions/authenticate",le="30000.0",} 1.0
keycloak_request_duration_bucket{method="POST",resource="/login-actions/authenticate",le="+Inf",} 1.0
keycloak_request_duration_count{method="POST",resource="/login-actions/authenticate",} 1.0
keycloak_request_duration_sum{method="POST",resource="/login-actions/authenticate",} 30.0
keycloak_request_duration_bucket{method="GET",resource="/admin/clients",le="50.0",} 13.0
keycloak_request_duration_bucket{method="GET",resource="/admin/clients",le="100.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin/clients",le="250.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin/clients",le="500.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin/clients",le="1000.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin/clients",le="2000.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin/clients",le="10000.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin/clients",le="30000.0",} 14.0
keycloak_request_duration_bucket{method="GET",resource="/admin/clients",le="+Inf",} 14.0
keycloak_request_duration_count{method="GET",resource="/admin/clients",} 14.0
keycloak_request_duration_sum{method="GET",resource="/admin/clients",} 136.0
keycloak_request_duration_bucket{method="GET",resource="",le="50.0",} 15.0
keycloak_request_duration_bucket{method="GET",resource="",le="100.0",} 15.0
keycloak_request_duration_bucket{method="GET",resource="",le="250.0",} 15.0
keycloak_request_duration_bucket{method="GET",resource="",le="500.0",} 15.0
keycloak_request_duration_bucket{method="GET",resource="",le="1000.0",} 15.0
keycloak_request_duration_bucket{method="GET",resource="",le="2000.0",} 15.0
keycloak_request_duration_bucket{method="GET",resource="",le="10000.0",} 15.0
keycloak_request_duration_bucket{method="GET",resource="",le="30000.0",} 15.0
keycloak_request_duration_bucket{method="GET",resource="",le="+Inf",} 15.0
keycloak_request_duration_count{method="GET",resource="",} 15.0
keycloak_request_duration_sum{method="GET",resource="",} 89.0
keycloak_request_duration_bucket{method="GET",resource="/protocol/openid-connect/auth",le="50.0",} 0.0
keycloak_request_duration_bucket{method="GET",resource="/protocol/openid-connect/auth",le="100.0",} 0.0
keycloak_request_duration_bucket{method="GET",resource="/protocol/openid-connect/auth",le="250.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/protocol/openid-connect/auth",le="500.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/protocol/openid-connect/auth",le="1000.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/protocol/openid-connect/auth",le="2000.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/protocol/openid-connect/auth",le="10000.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/protocol/openid-connect/auth",le="30000.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/protocol/openid-connect/auth",le="+Inf",} 2.0
keycloak_request_duration_count{method="GET",resource="/protocol/openid-connect/auth",} 2.0
keycloak_request_duration_sum{method="GET",resource="/protocol/openid-connect/auth",} 407.0
keycloak_request_duration_bucket{method="GET",resource="/admin/authentication",le="50.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/admin/authentication",le="100.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/admin/authentication",le="250.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/admin/authentication",le="500.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/admin/authentication",le="1000.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/admin/authentication",le="2000.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/admin/authentication",le="10000.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/admin/authentication",le="30000.0",} 2.0
keycloak_request_duration_bucket{method="GET",resource="/admin/authentication",le="+Inf",} 2.0
keycloak_request_duration_count{method="GET",resource="/admin/authentication",} 2.0
keycloak_request_duration_sum{method="GET",resource="/admin/authentication",} 5.0
```

## Why

This feature is require for better understanding the performance problems. 

## How

An additional "resource" field is added to the metrics.

## Verification Steps

1. Build the code
2. Copy the artifact into `$KC/standalone/deployments`
3. Make sure the metrics logger is turned on
4. Play with Keycloak a bit
5. Look at the following url: "$KC/auth/realms/master/metrics"

## Checklist:

- [x] Code has been tested locally by PR requester
- [ ] Changes have been successfully verified by another team member 

## Progress

- [x] Finished task

pb82 commented on src/main/java/org/jboss/aerogear/keycloak/metrics/ResourceExtractor.java:
```diff
@@ -0,0 +1,57 @@
+                    // resource ID.
+                    break;
+                }
+                if (realmIndex == -1 || i > realmIndex + 1) {
```
This seems to append everything to the string except the segment directly following the realm. How does it protect us from adding user ids to the result string?

The crucial thing here is to avoid adding information with lots of variation (like ids), because it will blow up the metrics size.

slaskawi replied:
I think you're right @pb82. Looking at our REST model (https://www.keycloak.org/docs-api/12.0/rest-api/index.html#_clients_resource) it seems we put most of the IDs (either client, user, flow or any other resource) in path params. I probably need another approach to get rid of them.

slaskawi replied:
Analyzing Matched URIs seems promissing:

```
13:28:52,766 INFO  [stdout] (default task-2) Matched URIs: [resources/89pe1/admin/logo-example/partials/client-list.html, resources]
13:30:35,495 INFO  [stdout] (default task-3) Matched URIs: [realms/master/account, realms/master/account, realms]
13:32:53,160 INFO  [stdout] (default task-10) Matched URIs: [realms/master/.well-known/openid-configuration, realms]
13:33:16,656 INFO  [stdout] (default task-10) Matched URIs: [realms/master/protocol/openid-connect/auth, realms/master/protocol/openid-connect, realms]
13:33:47,319 INFO  [stdout] (default task-10) Matched URIs: [realms/master/protocol/openid-connect/certs, realms/master/protocol/openid-connect, realms]
13:34:37,768 INFO  [stdout] (default task-10) Matched URIs: [realms/master/account/applications, realms/master/account, realms]
13:35:48,003 INFO  [stdout] (default task-11) Matched URIs: [admin/realms/master/client-scopes/ee5e5908-a0b6-43c8-b213-fb452f129a5e, admin/realms/master/client-scopes, admin/realms/master, admin/realms, admin]
13:36:24,642 INFO  [stdout] (default task-11) Matched URIs: [admin/realms/master/users/171753bc-8184-4989-929b-288fdc661b90, admin/realms/master/users, admin/realms/master, admin/realms, admin]
13:36:24,793 INFO  [stdout] (default task-11) Matched URIs: [admin/realms/master/attack-detection/brute-force/users/171753bc-8184-4989-929b-288fdc661b90, admin/realms/master/attack-detection, admin/realms/master, admin/realms, admin]
```

So I think the algorithm might take just the last two components from the vector. The only `if` that we'd have would be to "resources". If the resources are detected to be the last element - we'd just take the last item.

slaskawi replied:
@pb82 @maleck13 WDYT about this approach? Would it give us enough granularity?

maleck13 replied:
Sorry. I am not following what the metric would look like? 

In the case of 
```
[realms/master/protocol/openid-connect/certs, realms/master/protocol/openid-connect, realms]
```
would it look like

```
keycloak_request_duration_bucket{resource="realms/master/protocol/openid-connect/realms", method="POST"...}

```
Or something like 

```
keycloak_request_duration_bucket{resource="realms/master/protocol/openid-connect/", otherKey="realms", method="POST"...}

```

slaskawi replied:
Here's the output from my testing environment:

```
keycloak_request_duration_bucket{method="GET",resource="realms,realms/master/protocol/openid-connect",le="50.0",} 1.0
...
keycloak_request_duration_bucket{method="POST",resource="realms,realms/master/protocol/openid-connect",le="50.0",} 0.0
...
```

maleck13 replied:
wondering if splitting into two values would make it simpler to deal with from a query point of view. Might be a separate PR but might also be helpful to group by code.

```
keycloak_request_duration_bucket{method="GET",resource="realms", endpoint="realms/master/protocol/openid-connect",le="50.0",} 1.0
```

slaskawi replied:
That would be tricky as 99% of the request will fall into "realms" resource - everything that starts with `https://$KC/auth/realms`. Unfortunately this is the way the Keycloak JAX-RS resources are structured. The only one thing that you'd have separate is the admin endpoint. 

To sum it up - because of the way we use JAX-RS in Keycloak, it's very hard to do such things without modifying the code on the server (and attaching an SPI implementation).

maleck13 replied:
ok do we have any idea how many different values this would create? I guess it depends on the number of rest endpoints? Otherwise I think this is fine. Might be able to graph by doing a group by resource and also add specific alert around service accounts or openid-connect endpoint

slaskawi replied:
Yeah, it's hard to test all the endpoints at once. But in general you should have a distinction between the admin endpoints and the protocol ones (which is crucial for us).

pb82 commented on src/test/java/org/jboss/aerogear/keycloak/metrics/PrometheusExporterTest.java:
```diff
@@ -265,15 +265,18 @@ public void shouldCorrectlyRecordGenericAdminEvents() throws IOException {
     public void shouldCorrectlyRecordResponseErrors() throws IOException {
-        PrometheusExporter.instance().recordResponseError(500, "POST");
-        assertGenericMetric("keycloak_response_errors", 1, tuple("code", "500"), tuple("method", "POST"));
+        PrometheusExporter.instance().recordResponseError(500, "POST", "/auth/users");
```
can we add a test case that includes user ids to make sure they are not recorded?
(see https://github.com/aerogear/keycloak-metrics-spi/issues/64 )

slaskawi replied:
I changed the algorithm and I think this comment is not valid anymore.

slaskawi commented:
@maleck13 @pb82 I changed the algorithm to scrape the last two matched URIs. From my testing it looks good but I also added an "emergency turn off button".

pb82 commented:
@slaskawi thanks, i'll take another look

slaskawi commented:
@pb82 @maleck13 Can I can to review and merge this one? Once it's in, Peter, could I ask you for a release?

pb82 commented:
:eyes:

slaskawi commented:
@pb82 @maleck13 @abstractj Last chance to get this merged and onboarded into the production. After Wednesday, I'd say it's too late.
